src/district/server/smart_contracts.cljs

```diff
@@ -238,80 +238,95 @@
-                                                     (async/put! logs-ch {:err error :logs logs}))))
-                       logs-ch))]
-
-    ;; go chan by chan collecting events
-    (go-loop [all-logs []
-              [logs-ch & rest-logs] logs-chans]
-      (if logs-ch
-        (let [{:keys [:err :logs]} (async/<! logs-ch)
-              logs (map #(assoc % :err err) logs)]
-          ;; keep collecting
-          (recur (into all-logs logs) rest-logs))
-
-        ;; no more channels to read, sort and callback
-        (let [sorted-logs (cond->> (sort-by log-order-triplet all-logs)
-
-                            skip-log-indexes (remove (fn [l]
-                                                       (and (= (:block-number l) from-block)
-                                                            (skip-log-indexes [(:transaction-index l) (:log-index l)]))))
-                            true             transform-fn)]
-          (go-loop [logs sorted-logs]
-            (if (seq logs)
-              (do
-                (let [first-log (first logs)]
-
-                  (when (fn? callback)
-                    (doseq [res (try
-                                  (callback (:err first-log) (dissoc first-log :err))
-                                  (catch js/Error e
-                                    (when crash-on-event-fail?
-                                      (log/error "Server crash. Caused by event processing error with :crash-on-event-fail? true. Disable this flag to skip and continue.")
-                                      (.exit js/process 1))))]
-                      ;; if callback returns a promise or chan we block until it resolves
-                      (cond
-                        (satisfies? cljs.core.async.impl.protocols/ReadPort res)
-                        (<! res)
-
-                        (async-helpers/promise? res)
-                        (<! (async-helpers/promise->chan res))))))
-                (recur (rest logs)))
-
-              (when (fn? on-finish)
-                (on-finish sorted-logs)))))))))
+        next-chunk (fn [[from to]]
+                     (-> [(inc to) (+ to block-step)]
+                         (update 1 #(min % to-block))))
+        chunks (take-while (fn [[from to]] (<= from to-block))
+                           (iterate next-chunk
+                                    [from-block (min to-block (+ from-block (dec block-step)))]))
+        ch-from-blocks (async/to-chan! chunks)
+        ch-ordered-logs (async/chan 1)
+        sort-logs (fn [logs]
+                    (cond->> (sort-by log-order-triplet logs)
+                             skip-log-indexes (remove (fn [l]
+                                                        (and (= (:block-number l) from-block)
+                                                             (skip-log-indexes [(:transaction-index l) (:log-index l)]))))
+                             true transform-fn))
+        chunk-for-all-events (fn [[from to] ch]
```

Comment:
Another good candidate to extract and test separately - it has quite a bit of logic & branching, therefore number of things where unexpected behaviour (and bugs) can hide.